File: runme/mcgen/local/datan/datsrc/mincjg.c
#include <stdlib.h>
#include <limits.h>
#include "datsrc.h"

void LIBFUNC mincjg_(double P_T *x0, integer const P_T *n,
             integer const P_T *nred, integer const P_T *list,
             double (CBFUNC P_T *userfn)(double const P_T *,
             integer const P_T *), double P_T *fmin,
             double const P_T *epsiln, integer P_T *nstep)
{
    /* Table of constant values */

    double c_dm1 = -1.;
    double c_d0 = 0.;

    /* System generated locals */
    integer i__1;
    double d__1;

    /* Local variables */
    integer istep, lstep, nst;
    double a, gamma, s, fminl, eps;
    /*double gr[50], d[50], g[50], h[50];*/
    double *d, *g, *h, *gr;
    unsigned long amem;

    if ((amem = sizeof(double) * 4 * *n) > MAXALLOC) {
        MEMERROR("mincjg_ (too large)");
        goto f_0;
    }
    d = (double *)MEMALLOC(amem);
    if (!d) {
        MEMERROR("mincjg_");
        goto f_0;
    }
    g = d + *n; h = g + *n; gr = h + *n;

    /* Parameter adjustments */
    --list;
    --x0;

    /* Function Body */
    if (*nstep < 1) {
        *nstep = 1000;
    }
    eps = *epsiln;
    if (eps <= 0.) {
        eps = 1e-8;
    }
/* initialization : directions G and H identical to gradient at X0 */
    fminl = (*userfn)(&x0[1], n);
    if (*nred <= 0) {
/* all variables are fixed */
        *fmin = fminl;
        *nstep = 1;
        goto L40;
    }
    auxgrd_(&x0[1], gr, n, nred, &list[1], userfn);
    mtxmsv_(gr, g, &c_dm1, n);
    mtxcpv_(g, h, n);
/* start iteration */
    i__1 = *nstep;
    for (istep = 1; istep <= i__1; ++istep) {
        lstep = istep;
        if (istep > 1) {
            if ((d__1 = fminl - *fmin, abs(d__1)) < eps * abs(*fmin) + 1e-15) {
                *nstep = istep;
                goto L30;
            }
            fminl = *fmin;
        }
        nst = *nstep;
/* X0 is position of minimum along direction H */
        mindir_(&x0[1], h, n, userfn, fmin, &nst, &c_d0);
        if (nst < 0) {
            goto L20;
        }
        auxgrd_(&x0[1], gr, n, nred, &list[1], userfn);
/* GR is gradient at X0 */
/* compute next set of directions G and H */
        mtxmsv_(gr, gr, &c_dm1, n);
        mtxsbv_(gr, g, d, n);
        mtxdot_(d, gr, &s, n);
        mtxdot_(g, g, &a, n);
        if (a == 0.) {
            goto L30;
        }
        gamma = s / a;
        mtxmsv_(h, h, &gamma, n);
        mtxadv_(gr, h, h, n);
        mtxcpv_(gr, g, n);
    }
L20:
    *nstep = -1;
    goto L40;
L30:
    *nstep = lstep;
L40:
    MEMFREE((void *)d);
f_0:
    return;
} /* mincjg_ */

File: runme/mcgen/local/datan/datsrc/mincjg.f
      SUBROUTINE MINCJG(X0,N,NRED,LIST,USERFN,FMIN,EPSILN,NSTEP)
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      PARAMETER(MAXN=50)
      DIMENSION X0(N),GR(MAXN),G(MAXN),H(MAXN),D(MAXN),LIST(N)
      PARAMETER(MAXSTP=1000,EPSDEF=1.D-8,T=1.D-15,ZERO=0.D0,
     +ONE=1.D0)
      EXTERNAL USERFN
      IF(NSTEP.LT.1) NSTEP=MAXSTP
      EPS=EPSILN
      IF(EPS.LE.ZERO) EPS=EPSDEF
C initialization : directions G and H identical to gradient at X0
      FMINL=USERFN(X0,N)
      IF(NRED.LE.0) THEN
C all variables are fixed
        FMIN=FMINL
        NSTEP=1
        GO TO 40
      END IF
      CALL AUXGRD(X0,GR,N,NRED,LIST,USERFN)
      CALL MTXMSV(GR,G,-ONE,N)
      CALL MTXCPV(G,H,N)
C start iteration
      DO 10 ISTEP=1,NSTEP
        LSTEP=ISTEP
        IF(ISTEP.GT.1) THEN
          IF(ABS(FMINL-FMIN) .LT. EPS*ABS(FMIN)+T) THEN
            NSTEP=ISTEP
            GO TO 30
          END IF
          FMINL=FMIN
        END IF
        NST=NSTEP
C X0 is position of minimum along direction H
        CALL MINDIR(X0,H,N,USERFN,FMIN,NST,ZERO)
        IF(NST.LT.0) GO TO 20
        CALL AUXGRD(X0,GR,N,NRED,LIST,USERFN)
C GR is gradient at X0
C compute next set of directions G and H
        CALL MTXMSV(GR,GR,-ONE,N)
        CALL MTXSBV(GR,G,D,N)
        CALL MTXDOT(D,GR,S,N)
        CALL MTXDOT(G,G,A,N)
        IF(A.EQ.ZERO) GO TO 30
        GAMMA=S/A
        CALL MTXMSV(H,H,GAMMA,N)
        CALL MTXADV(GR,H,H,N)
        CALL MTXCPV(GR,G,N)
   10 CONTINUE
   20 NSTEP=-1
      GO TO 40
   30 NSTEP=LSTEP
   40 CONTINUE
      END
